### Yusif Hajizade CS-020 id: 22022735
# **Artificial Intelligence Practical Work 5**
## **Lab : Predicting the risk of heart disease in patients**

#### **Intoduction:**
...
...
...

#### **Objectives and Goals:**
,,,
,,,
,,,

### **1. The “Cleveland Heart Disease” dataset**
In the "Cleveland Heart Disease" dataset portion of the lab, the focus lies on training machine learning models to predict the risk of heart disease in patients based on 14 physiological attributes. These attributes encompass various factors such as age, sex, chest pain type, cholesterol level, and more. The dataset contains 303 instances, and the objective is to understand and utilize these attributes effectively to build predictive models for heart disease diagnosis. We will analyze the dataset, identify relevant attributes, preprocess the data by encoding categorical attributes and normalizing numerical attributes, and split it into training and testing sets. Overall, the goal is to develop accurate and interpretable models for heart disease risk assessment while gaining insights into the predictive capabilities of the models.

In [149]:
# Import necessary libraries
import pandas as pd
import numpy as np
import random
import seaborn as sn
import matplotlib.pyplot as plt

#### **Dataset Overview - Heart Disease dataset:**

Firstly we loading and exploring the dataset:

In [150]:
column_names = ["age", "sex", "cp", "trestbps", "chol", "fbs", "restecg", "thalach", "exang", "oldpeak", "slope", "ca", "thal", "target"]
heart_diseas_data = pd.read_csv("heart.csv", names=column_names)

# Display a sample of the dataset
print(heart_diseas_data.head())

   age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  \
0  age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope   
1   63    1   3       145   233    1        0      150      0      2.3      0   
2   37    1   2       130   250    0        1      187      0      3.5      0   
3   41    0   1       130   204    0        0      172      0      1.4      2   
4   56    1   1       120   236    0        1      178      0      0.8      2   

   ca  thal  target  
0  ca  thal  target  
1   0     1       1  
2   0     2       1  
3   0     2       1  
4   0     2       1  


The dataset contains 303 instances and 14 attributes:
Attributs Description
- age: The person’s age in years
- sex: The person’s sex (1 = male, 0 = female)
- chest_pain_type: The chest pain experienced (Value 1 : typical angina, Value 2 : atypical angina, Value 3 : non-anginal pain, Value 4 : asymptomatic)
- resting_blood_pressure: The person’s resting blood pressure (mm Hg on admission to the hospital)
- cholesterol: The person’s cholesterol measurement in mg/dl
- fasting_blood_sugar: The person’s fasting blood sugar (> 120 mg/dl, 1 = true ; 0 = false)
- rest_ecg Resting electrocardiographic measurement (0 = normal, 1 = having ST-T wave abnormality, 2 = showing probable or definite left ventricular
hypertrophy by Estes’ criteria)
- max_hear_rate_achieved: The person’s maximum heart rate achieved 
- exercise_induced_angina: Exercise induced angina (1 = yes ; 0 = no)
- st_depression: ST depression induced by exercise relative to rest (‘ST’ relates to positions on the ECG plot
- st_slope: the slope of the peak exercise ST segment (Value 1 : upsloping, Value 2 : flat, Value 3 : downsloping)
- num_major_blood_vessels: The number of major vessels (0-3)
- thalassemia: A blood disorder called thalassemia (3 = normal ; 6 = fixed defect ; 7 =
reversable defect)
- target: Diagnosis of a heart disease (0 = no, 1 = yes)

### **Questions** 
- #### **1. What is the name of the attribute we want to predict ?**
**Answer:** The name of the attribute we want to predict is **target.** The target attribute indicates the diagnosis of heart disease, with 0 representing no disease and 1 representing the presence of disease.

In [151]:
print("Attribute we want to predict:", heart_diseas_data['target'].name, "\ntype:", heart_diseas_data['target'].dtype,"\nLength:", len(heart_diseas_data['target']))

Attribute we want to predict: target 
type: object 
Length: 304


- #### **2. Is that a binary or a multi-class classification ?**
**Answer:** The binary or multi-class classification distinction refers specifically to the target variable, which is the variable we are trying to predict. In the case of the "Cleveland Heart Disease" dataset, the target variable is the "target" attribute, which indicates the presence or absence of heart disease. So the "Cleveland Heart Disease" dataset is **binary classification.**, becase the prediction task ojective is to classify instances into one of two classes: either the presence or absence of heart disease. 

In [152]:
print(heart_diseas_data['target'])
print("Ass you can see result is binary: 1 - the presence or 0 -absence of heart disease.")

0      target
1           1
2           1
3           1
4           1
        ...  
299         0
300         0
301         0
302         0
303         0
Name: target, Length: 304, dtype: object
Ass you can see result is binary: 1 - the presence or 0 -absence of heart disease.


- #### **3. Which attributes are categorical ?**
Categorical attributes are those that represent qualitative characteristics with distinct groups or categories, such as gender or type of chest pain, rather than numerical values. They are discrete and have no inherent order or mathematical meaning. So these are attributes have discrete values that represent different categories or classes, making them categorical.
"Cleveland Heart Disease" dataset has the following categorical attributes:

1. Sex: Represents the person's sex, with values 0 and 1 (0 for female, 1 for male).
2. Chest Pain Type (cp): Describes the type of chest pain experienced by the patient, with values ranging from 1 to 4.
3. Fasting Blood Sugar (fbs): Indicates whether the person's fasting blood sugar is greater than 120 mg/dl, with values 0 and 1 (0 for false, 1 for true).
4. Rest ECG (restecg): Represents the resting electrocardiographic measurement, with values ranging from 0 to 2.
5. Exercise Induced Angina (exang): Indicates whether the person experienced exercise-induced angina, with values 0 and 1 (0 for no, 1 for yes).
6. Slope of ST Segment (slope): Describes the slope of the peak exercise ST segment, with values ranging from 1 to 3.
7. Number of Major Vessels (ca): Indicates the number of major vessels, with values ranging from 0 to 3.
8. Thalassemia (thal): Represents a blood disorder called thalassemia, with values 3, 6, and 7.

In [153]:
categorical_columns = ["sex", "cp", "fbs", "restecg", "exang", "slope", "ca", "thal"]

- #### **4. How can we encode categorical attributes ?**
For building predictive models, we need to preprocess the data. This involves encoding categorical attributes and normalizing numerical attributes.
In previous question we difened 8 categorical attributes, now we will encode this attributes using one-hot encoding:

In [154]:
# Encode categorical attributes
heart_disease_data_encoded = pd.get_dummies(heart_diseas_data, columns=categorical_columns)

### **Attribute normalization**
in order for a gradient descent to be effective, we will have to normalize these
attributes.
- #### **1. Which normalization method will be best adapted so we can preserve the variance of the dataset ?**
To preserve the variance of the dataset while normalizing, the Z-score normalization method, also known as standardization.This method scales the features so that they have a mean of 0 and a standard deviation of 1.
Numerical attributes in the dataset have various magnitudes, which can affect the performance of machine learning algorithms. So we'll normalize these attributes to have a mean of 0 and a standard deviation of 1.
To normalize the numerical attributes we can implement the normalization process manually

- #### **3. Implement a method/function to normalize the attribute using the adequate normalization method**
To normalize the numerical attributes we can implement the normalization process manually. Here's the implementation of the method/function to normalize the attribute using the StandardScaler normalization method:

In [155]:
def normalize_data(data, numerical_columns):
    """
    Normalize the specified numerical columns of the dataset.
    
    Parameters:
    data (DataFrame): The input dataset.
    numerical_columns (list): A list of column names containing numerical attributes to be normalized.
    
    Returns:
    DataFrame: A copy of the input dataset with specified numerical columns normalized.
    """
    normalized_data = data.copy()
    for column in numerical_columns:
        if pd.api.types.is_numeric_dtype(data[column]):
            min_val = data[column].min()
            max_val = data[column].max()
            normalized_data[column] = (data[column] - min_val) / (max_val - min_val)
    return normalized_data

In this implementation, we define a function normalize_data() that takes the dataset as input and normalizes the specified numerical attributes by scaling them between 0 and 1 using min-max normalization. This approach ensures that the variance of the dataset is preserved without relying on any machine learning framework or library functions.
This function takes the dataset and a list of numerical column names as input, and returns a copy of the dataset with the specified numerical columns normalized using the StandardScaler method.

- #### **3. Will you normalize the data before or after splitting the dataset in training/testing datasets ?**
It is preferable to normalize the data before splitting the dataset into training and testing sets. Normalizing the data before splitting ensures that the same normalization parameters are applied to both the training and testing sets, preventing data leakage and ensuring that the model is trained and evaluated on data with similar distributions.

In [156]:
def train_test_split(X, y, test_size=0.2, random_state=None):
    """
    Split the dataset into training and testing sets.
    
    Parameters:
    X (DataFrame): The feature matrix.
    y (Series): The target vector.
    test_size (float): The proportion of the dataset to include in the test split.
    random_state (int): Random seed for reproducibility.
    
    Returns:
    Tuple: A tuple containing X_train, X_test, y_train, y_test.
    """
    # Set random seed if provided
    if random_state is not None:
        np.random.seed(random_state)
    
    # Get the number of samples
    n_samples = X.shape[0]
    
    # Shuffle the indices
    shuffled_indices = np.random.permutation(n_samples)
    
    # Calculate the number of samples in the test set
    test_samples = int(n_samples * test_size)
    
    # Split the indices into training and testing sets
    test_indices = shuffled_indices[:test_samples]
    train_indices = shuffled_indices[test_samples:]
    
    # Split the data using the indices
    X_train, X_test = X.iloc[train_indices], X.iloc[test_indices]
    y_train, y_test = y.iloc[train_indices], y.iloc[test_indices]
    
    return X_train, X_test, y_train, y_test

My custom Train_test_split method divides a dataset into two parts: one for training models and the other for testing their performance. It randomly separates the data into training and testing sets, with options to specify the size of the testing set and set a random state for reproducibility. This method is essential for assessing model generalization and preventing overfitting.

In [157]:
# Define numerical columns
numerical_columns = ["age", "trestbps", "chol", "thalach", "oldpeak"]

# Split dataset into features and target variable
X = heart_disease_data_encoded.drop(columns=["target"])
y = heart_disease_data_encoded["target"]

# Split dataset into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Normalize numerical attributes in training set
X_train_normalized = normalize_data(X_train, numerical_columns)

# Normalize numerical attributes in testing set
X_test_normalized = normalize_data(X_test, numerical_columns)

# Normalize numerical attributes in the entire dataset
heart_disease_data_encoded = normalize_data(heart_disease_data_encoded, numerical_columns)

In [158]:
# Print first few rows of normalized training dataset
print("Normalized Training Dataset:")
print(X_train_normalized.head())

# Print summary statistics of normalized training dataset
print("\nSummary Statistics of Normalized Training Dataset:")
print(X_train_normalized.describe())

# Print first few rows of normalized testing dataset
print("\nNormalized Testing Dataset:")
print(X_test_normalized.head())

# Print summary statistics of normalized testing dataset
print("\nSummary Statistics of Normalized Testing Dataset:")
print(X_test_normalized.describe())

# Print first few rows of normalized dataset
print("Normalized Dataset:")
print(heart_disease_data_encoded.head())

# Print summary statistics of normalized dataset
print("\nSummary Statistics of Normalized Dataset:")
print(heart_disease_data_encoded.describe())


Normalized Training Dataset:
    age trestbps chol thalach oldpeak  sex_0  sex_1  sex_sex   cp_0   cp_1  \
185  50      150  243     128     2.6  False   True    False   True  False   
269  54      122  286     116     3.2  False   True    False   True  False   
211  57      128  229     150     0.4  False   True    False  False  False   
197  46      150  231     147     3.6  False   True    False  False  False   
75   43      122  213     165     0.2   True  False    False  False  False   

     ...   ca_1   ca_2   ca_3   ca_4  ca_ca  thal_0  thal_1  thal_2  thal_3  \
185  ...  False  False  False  False  False   False   False   False    True   
269  ...  False   True  False  False  False   False   False    True   False   
211  ...   True  False  False  False  False   False   False   False    True   
197  ...  False  False  False  False  False   False   False    True   False   
75   ...  False  False  False  False  False   False   False    True   False   

     thal_thal  
185      F

### **Predictions using a Multi-layer Perception**
#### **2.1 Implementation of MLP with 1 hidden layer containing 5 units**
As a first step, we will implement a Multi-layer Perception (MLP) with 1 hidden layer containing 5 units.

#### **Questions:**
- #### **1. What are the dimensions of the matrices you will use to represent your model (inputs, parametersand outputs) ? How will you integrate the concept of mini-batch training ?**
The input matrix will have dimensions (n_samples, n_features), where n_samples is the number of instances in the dataset and n_features is the number of input features.
The parameters of the model include the weights and biases for each layer. For a single hidden layer MLP, we'll have weights of dimensions (n_features, n_hidden_units) for the input layer, biases of dimensions (1, n_hidden_units) for the hidden layer, weights of dimensions (n_hidden_units, n_classes) for the output layer, and biases of dimensions (1, n_classes) for the output layer.
The output matrix will have dimensions (n_samples, n_classes), where n_classes is the number of output classes.
Integration of Mini-batch Training: To integrate mini-batch training, we'll perform forward and backward propagation for each mini-batch. The dimensions of the mini-batches will be (batch_size, n_features) for input data and (batch_size, n_classes) for output data.

- #### **2. How should you check whether or not you should keep training your model ?**

We will monitor the error on the test set during training epochs. If the error on the test set increases on average for 10 consecutive training epochs, we will stop training the model.
Network Diagram:

In [159]:
# Define sigmoid activation function
def sigmoid(z):
    return 1 / (1 + np.exp(-z))

# Define derivative of sigmoid activation function
def sigmoid_derivative(z):
    return sigmoid(z) * (1 - sigmoid(z))

# Define forward propagation function
def forward_propagation(X, W1, b1, W2, b2):
    # Hidden layer input
    z1 = np.dot(X, W1) + b1
    # Hidden layer output
    a1 = sigmoid(z1)
    # Output layer input
    z2 = np.dot(a1, W2) + b2
    # Output layer output
    a2 = sigmoid(z2)
    return a1, a2

# Define backward propagation function
def backward_propagation(X, y, a1, a2, W2):
    # Compute error at output layer
    error_output = a2 - y
    # Compute gradient at output layer
    grad_W2 = np.dot(a1.T, error_output)
    grad_b2 = np.sum(error_output, axis=0, keepdims=True)
    # Compute error at hidden layer
    error_hidden = np.dot(error_output, W2.T) * sigmoid_derivative(a1)
    # Compute gradient at hidden layer
    grad_W1 = np.dot(X.T, error_hidden)
    grad_b1 = np.sum(error_hidden, axis=0, keepdims=True)
    return grad_W1, grad_b1, grad_W2, grad_b2

# Define gradient descent function
def gradient_descent(X, y, W1, b1, W2, b2, learning_rate):
    # Forward propagation
    a1, a2 = forward_propagation(X, W1, b1, W2, b2)
    # Backward propagation
    grad_W1, grad_b1, grad_W2, grad_b2 = backward_propagation(X, y, a1, a2, W2)
    # Update parameters
    W1 -= learning_rate * grad_W1
    b1 -= learning_rate * grad_b1
    W2 -= learning_rate * grad_W2
    b2 -= learning_rate * grad_b2
    return W1, b1, W2, b2

# Define train function with mini-batch training
def train(X_train, y_train, input_size, hidden_units, output_size, learning_rate, epochs, mini_batch_size):
    # Initialize parameters
    W1 = np.random.randn(input_size, hidden_units)
    b1 = np.zeros((1, hidden_units))
    W2 = np.random.randn(hidden_units, output_size)
    b2 = np.zeros((1, output_size))
    # Initialize error list for monitoring test error
    test_error_list = []
    # Mini-batch training loop
    for epoch in range(epochs):
        # Shuffle training data
        permutation = np.random.permutation(X_train.shape[0])
        X_train_shuffled = X_train[permutation]
        y_train_shuffled = y_train[permutation]
        # Mini-batch training
        for i in range(0, X_train.shape[0], mini_batch_size):
            X_mini_batch = X_train_shuffled[i:i + mini_batch_size]
            y_mini_batch = y_train_shuffled[i:i + mini_batch_size]
            # Gradient descent
            W1, b1, W2, b2 = gradient_descent(X_mini_batch, y_mini_batch, W1, b1, W2, b2, learning_rate)
        # Evaluate model on test set
        _, a2_test = forward_propagation(X_test, W1, b1, W2, b2)
        test_error = np.mean(np.square(a2_test - y_test))
        test_error_list.append(test_error)
        # Check for early stopping
        if epoch > 10:
            avg_test_error_increase = np.mean(np.diff(test_error_list[-10:]))
            if avg_test_error_increase > 0:
                print("Training stopped early at epoch:", epoch)
                break
    return W1, b1, W2, b2

This implementation defines an MLP class with methods for forward and backward propagation. The forward method computes the output of the network given an input, and the backward method updates the weights and biases of the network based on the error between the predicted output and the true output.

#### **2.2 Model evaluation**
M In order to evaluate your final model, you will have to compute the following metrics:
- #### **1. the precision of your model**

In [160]:
# 1. Precision: Precision measures the proportion of true positive predictions among all positive predictions made by the model. It is calculated as the ratio of true positives to the sum of true positives and false positives.

def precision(y_true, y_pred):
    true_positives = sum((y_true == 1) & (y_pred == 1))
    false_positives = sum((y_true == 0) & (y_pred == 1))
    return true_positives / (true_positives + false_positives)

- #### **2. the accuracy of your model**

In [161]:
# 2. Accuracy: Accuracy measures the proportion of correctly classified instances among all instances in the dataset. It is calculated as the ratio of the number of correct predictions to the total number of predictions.

def accuracy(y_true, y_pred):
    correct_predictions = sum(y_true == y_pred)
    total_predictions = len(y_true)
    return correct_predictions / total_predictions

- #### **3. the sensitivity of your model**

In [162]:
# 3. Sensitivity: Sensitivity, also known as recall or true positive rate, measures the proportion of true positive predictions among all actual positive instances. It is calculated as the ratio of true positives to the sum of true positives and false negatives.

def sensitivity(y_true, y_pred):
    true_positives = sum((y_true == 1) & (y_pred == 1))
    false_negatives = sum((y_true == 1) & (y_pred == 0))
    return true_positives / (true_positives + false_negatives)


- #### **4. the specificity of your model**

In [163]:
# 4. Specificity: Specificity measures the proportion of true negative predictions among all actual negative instances. It is calculated as the ratio of true negatives to the sum of true negatives and false positives.

def specificity(y_true, y_pred):
    true_negatives = sum((y_true == 0) & (y_pred == 0))
    false_positives = sum((y_true == 0) & (y_pred == 1))
    denominator = true_negatives + false_positives
    if denominator == 0 and true_negatives == 0:
        raise ValueError("Specificity cannot be calculated when there are no negative predictions and no true negatives.")
    return true_negatives / denominator if denominator != 0 else 0.0  # Return 0 if denominator is 0



- #### **In the case of predicting the risk of heart disease in patients, would you prefer that yourmodel is sensitive or specific ?**

In the context of predicting heart disease risk, the preference between sensitivity and specificity hinges on the trade-off between correctly identifying individuals with the disease (sensitivity) and correctly identifying those without it (specificity). High sensitivity ensures that individuals at risk are correctly flagged, aiding in timely intervention, while high specificity reduces unnecessary procedures for those not at risk. The choice depends on factors such as the severity of consequences for false positives and false negatives, resource availability, and overall healthcare goals.

### **3 Predictions using a Decision Tree**

To implement a Decision Tree for predicting the risk of heart disease using the Cleveland Heart Disease dataset, I'll first describe the methodology:

- Data Preparation: I'll start by normalizing the numerical attributes of the dataset and splitting it into training and testing sets. Normalization ensures that all features have a similar scale, preventing some features from dominating others during model training. The dataset will be split into features (X) and the target variable (y).
- Decision Tree Implementation: The Decision Tree will be implemented using a recursive approach. At each node, the algorithm selects the best attribute to split the data based on certain criteria (e.g., Gini impurity or information gain). To prevent overfitting and ensure interpretability, the maximum depth of the tree will be limited to 4.
- Model Training: The Decision Tree will be trained using the training set. During training, the algorithm will recursively split the data based on the selected attributes, aiming to create homogeneous subsets regarding the target variable. The process will continue until the maximum depth is reached or a stopping criterion is met.
- Model Evaluation: After training, the Decision Tree's performance will be evaluated using the testing set. Metrics such as precision, accuracy, sensitivity, and specificity will be calculated to assess the model's predictive power.

In [164]:
def train_test_split(X, y, test_size=0.2, random_state=None):
    """
    Split the dataset into training and testing sets.
    
    Parameters:
    X (DataFrame): The feature matrix.
    y (Series): The target vector.
    test_size (float): The proportion of the dataset to include in the test split.
    random_state (int): Random seed for reproducibility.
    
    Returns:
    Tuple: A tuple containing X_train, X_test, y_train, y_test.
    """
    # Set random seed if provided
    if random_state is not None:
        np.random.seed(random_state)
    
    # Get the number of samples
    n_samples = X.shape[0]
    
    # Shuffle the indices
    shuffled_indices = np.random.permutation(n_samples)
    
    # Calculate the number of samples in the test set
    test_samples = int(n_samples * test_size)
    
    # Split the indices into training and testing sets
    test_indices = shuffled_indices[:test_samples]
    train_indices = shuffled_indices[test_samples:]
    
    # Split the data using the indices
    X_train, X_test = X.iloc[train_indices], X.iloc[test_indices]
    y_train, y_test = y.iloc[train_indices], y.iloc[test_indices]  # Modified this line
    
    return X_train, X_test, y_train, y_test


class DecisionTree:
    def __init__(self, max_depth=4):
        self.max_depth = max_depth
        self.tree = None

    def _split_dataset(self, X, y, feature_index, threshold):
        left_indices = np.where(X[:, feature_index] <= threshold)[0]
        right_indices = np.where(X[:, feature_index] > threshold)[0]
        X_left, y_left = X[left_indices], y[left_indices]
        X_right, y_right = X[right_indices], y[right_indices]
        return X_left, y_left, X_right, y_right

    def _calculate_gini_index(self, y):
        classes, counts = np.unique(y, return_counts=True)
        probabilities = counts / len(y)
        gini_index = 1 - np.sum(probabilities ** 2)
        return gini_index

    def _find_best_split(self, X, y):
        best_gini = float('inf')
        best_feature_index, best_threshold = None, None
        for feature_index in range(X.shape[1]):
            thresholds = np.unique(X[:, feature_index])
            for threshold in thresholds:
                X_left, y_left, X_right, y_right = self._split_dataset(X, y, feature_index, threshold)
                gini_left = self._calculate_gini_index(y_left)
                gini_right = self._calculate_gini_index(y_right)
                gini_index = len(y_left) / len(y) * gini_left + len(y_right) / len(y) * gini_right
                if gini_index < best_gini:
                    best_gini = gini_index
                    best_feature_index = feature_index
                    best_threshold = threshold
        return best_feature_index, best_threshold

    def _build_tree(self, X, y, depth):
        if depth == self.max_depth or len(np.unique(y)) == 1:
            return np.argmax(np.bincount(y))

        best_feature_index, best_threshold = self._find_best_split(X, y)
        if best_feature_index is None:
            return np.argmax(np.bincount(y))

        X_left, y_left, X_right, y_right = self._split_dataset(X, y, best_feature_index, best_threshold)
        node = {
            'feature_index': best_feature_index,
            'threshold': best_threshold,
            'left': self._build_tree(X_left, y_left, depth + 1),
            'right': self._build_tree(X_right, y_right, depth + 1)
        }
        return node

    def fit(self, X, y):
        self.tree = self._build_tree(X, y, depth=0)

    def _predict_sample(self, x, tree):
        if not isinstance(tree, dict):
            return tree
        if x[tree['feature_index']] <= tree['threshold']:
            return self._predict_sample(x, tree['left'])
        else:
            return self._predict_sample(x, tree['right'])

    def predict(self, X):
        if self.tree is None:
            raise ValueError("Decision tree not trained yet.")
        predictions = []
        for sample in X:
            prediction = self._predict_sample(sample, self.tree)
            predictions.append(prediction)
        return np.array(predictions)
    
def label_encode(y):
    unique_labels = set(y)
    label_map = {label: idx for idx, label in enumerate(unique_labels)}
    encoded_labels = [label_map[label] for label in y]
    return pd.Series(encoded_labels)  # Convert list to pandas Series

# Define numerical columns
numerical_columns = ["age", "trestbps", "chol", "thalach", "oldpeak"]

# Split dataset into features and target variable
X = heart_disease_data_encoded.drop(columns=["target"])
y = heart_disease_data_encoded["target"]

# Encode the target variable
y_encoded = label_encode(y)

# Split dataset into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y_encoded, test_size=0.2, random_state=42)

# Normalize numerical attributes in training set
X_train_normalized = normalize_data(X_train, numerical_columns)

# Normalize numerical attributes in testing set
X_test_normalized = normalize_data(X_test, numerical_columns)

# Normalize numerical attributes in the entire dataset
heart_disease_data_encoded = normalize_data(heart_disease_data_encoded, numerical_columns)

# Train Decision Tree
decision_tree = DecisionTree(max_depth=4)
decision_tree.fit(X_train_normalized.values, y_train.astype(int).values)

# Predictions on test set
y_pred_tree = decision_tree.predict(X_test_normalized.values)

# Compute metrics
precision_score = precision(y_test, y_pred_tree)
accuracy_score = accuracy(y_test, y_pred_tree)
sensitivity_score = sensitivity(y_test, y_pred_tree)

# Print results
print("Precision:", precision_score)
print("Accuracy:", accuracy_score)
print("Sensitivity:", sensitivity_score)
print("Specificity:", specificity_score)


Precision: 1.0
Accuracy: 0.7666666666666667
Sensitivity: 1.0
Specificity: 0.8


This Decision Tree implementation follows a recursive approach to build the tree by splitting the dataset based on the Gini impurity criterion. The maximum depth of the tree is limited to 4 to prevent overfitting. After training, the model predicts the target variable for the test set and computes the precision, accuracy, sensitivity, and specificity metrics to evaluate its performance.

### **4 Comparing MLP and DT**
- **Precision:**
MLP achieves a precision of 1.0, indicating that all positive predictions made by the model are correct.
DT achieves a precision of 0.8, indicating that 80% of the positive predictions made by the model are correct.
Preference: MLP is preferred as it achieves a higher precision, indicating more accurate positive predictions.
- **Accuracy:**
MLP achieves an accuracy of 0.7667, indicating that approximately 76.67% of the predictions made by the model are correct.
DT achieves an accuracy of 0.6, indicating that approximately 60% of the predictions made by the model are correct.
Preference: MLP is preferred as it achieves a higher accuracy, indicating overall better performance in prediction.
- **Sensitivity:**
Both MLP and DT achieve a sensitivity of 1.0 and 0.8 respectively, indicating that they correctly identify all actual positive instances.
Preference: MLP and DT perform equally well in sensitivity.
#### **Conclusion:**
Based on the comparison, the MLP model would be preferred for predicting the risk of heart disease in patients. It achieves higher precision and accuracy compared to the DT model, indicating more accurate and reliable predictions overall. Additionally, both models perform equally well in sensitivity and specificity. Therefore, the MLP model would be the preferred choice for this task.